### PR TITLE
Dashboard Tab entity

### PR DESCRIPTION
**Pull request by qnkhuat**
BE for milestone 1 of https://github.com/metabase/metabase/issues/29502

This PR adds a new table `report_dashboardtab` that will be used to store tabs info for a dashboard.
To know how it's going to be used, see https://github.com/metabase/metabase/pull/29861

[Eng doc](https://www.notion.so/metabase/Engineering-Notes-Data-Model-Plan-92672ca13d9c4d45921df269bf13a657?pvs=4)

This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/29802)

### CODE REVIEW

File: enterprise/backend/src/metabase_enterprise/serialization/v2/seed_entity_ids.clj [hunk 1]
```diff
@@ -34,13 +34,14 @@
...
   []
-  (into {} (for [model (descendants :toucan1/model)
-                 :when (models/model? model)
-                 :let  [table-name (some-> model t2/table-name name)]
-                 :when table-name
-                 ;; ignore any models defined in test namespaces.
-                 :when (not (str/includes? (namespace model) "test"))]
-             [table-name model])))
+  (into {}
+        (for [model (concat (descendants :toucan1/model) (descendants :metabase/models))
```
Comment (qnkhuat) [inline enterprise/backend/src/metabase_enterprise/serialization/v2/seed_entity_ids.clj:38]: see this for background on :metabase/models https://github.com/metabase/metabase/pull/29785#pullrequestreview-1375413845

File: resources/migrations/000_migrations.yaml [hunk 2]
```diff
@@ -14378,6 +14378,93 @@ databaseChangeLog:
...
+  - changeSet:
+      id: v47.00-007
+      author: qnkhuat
+      comment: Added 0.47.0 -- add report_dashboardcard.dashboardtab_id
+      changes:
+        - addColumn:
+            tableName: report_dashboardcard
+            columns:
+              - column:
+                  name: dashboardtab_id
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply resources/migrations/000_migrations.yaml:14450]: I took this because we have `dashboardcard_series.dashboardcard_id`

**Review by calherries — CHANGES_REQUESTED** [review]
> Reviewed 6 of 10 files at r1.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29802)* status: 6 of 10 files reviewed, 5 unresolved discussions (waiting on @qnkhuat)

___
*[`enterprise/backend/src/metabase_enterprise/serialization/v2/seed_entity_ids.clj` line 38 at r1](https://reviewable.io/reviews/metabase/metabase/29802#gh-1162447821:-NSjVoKb-iyalRu6l96t:be1jqcu) ([raw file](https://github.com/metabase/metabase/blob/a659d80b64a851d1e03e0f8ca95b9ed99e5ca8e6/enterprise/backend/src/metabase_enterprise/serialization/v2/seed_entity_ids.clj#L38)):*
<details><summary><i>Previously, qnkhuat (Ngoc Khuat) wrote…</i></summary><blockquote>

see this for background on why we use `:metabase/models` https://github.com/metabase/metabase/pull/29785#pullrequestreview-1375413845
</blockquote></details>

Shouldn't it be `:metabase/model`?
___
*[`resources/migrations/000_migrations.yaml` line 14387 at r1](https://reviewable.io/reviews/metabase/metabase/29802#-NSjTWrI5c-Ui-gVa1wF:-NSjTWrI5c-Ui-gVa1wG:biphyvs) ([raw file](https://github.com/metabase/metabase/blob/a659d80b64a851d1e03e0f8ca95b9ed99e5ca8e6/resources/migrations/000_migrations.yaml#L14387)):*
> ```YAML
>       changes:
>         - createTable:
>             tableName: report_dashboardtab
> ```

Why not `dashboard_tab`? There's already no consistency between tables with `report_` in front and those without, so I would prefer to minimize the already small number of tables with `report_` in front and have most tables not have it.
___
*[`src/metabase/models/dashboard.clj` line 58 at r1](https://reviewable.io/reviews/metabase/metabase/29802#-NSjUweO5qOG2ngLAw6w:-NSjUweO5qOG2ngLAw6x:b-tzccb2) ([raw file](https://github.com/metabase/metabase/blob/a659d80b64a851d1e03e0f8ca95b9ed99e5ca8e6/src/metabase/models/dashboard.clj#L58)):*
> ```Clojure
>               :order-by  [[:dashcard.created_at :asc]]}))
> ```

Extra whitespace
___
*[`src/metabase/models/dashboard.clj` line 61 at r1](https://reviewable.io/reviews/metabase/metabase/29802#-NSjV1ez5iLa1GGMAqGw:-NSjV1f-D00uKwUUWelY:undefined) ([raw file](https://github.com/metabase/metabase/blob/a659d80b64a851d1e03e0f8ca95b9ed99e5ca8e6/src/metabase/models/dashboard.clj#L61)):*
> ```Clojure
> (mi/define-simple-hydration-method ordered-tabs
>   :ordered_tabs
>   "Return the DashboardCards associated with `dashboard`, in the order they were created."
> ```





_Suggestion:_
```Clojure
  "Return the ordered DashboardTabs associated with `dashboard-or-id`."
```

**Review by qnkhuat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29802)* status: 4 of 10 files reviewed, 5 unresolved discussions (waiting on @calherries)

___
*[`resources/migrations/000_migrations.yaml` line 14387 at r1](https://reviewable.io/reviews/metabase/metabase/29802#-NSjTWrI5c-Ui-gVa1wF:-NSjqKiPDG00wKviQTCq:b-23miyc) ([raw file](https://github.com/metabase/metabase/blob/a659d80b64a851d1e03e0f8ca95b9ed99e5ca8e6/resources/migrations/000_migrations.yaml#L14387)):*
<details><summary><i>Previously, calherries (Cal Herries) wrote…</i></summary><blockquote>

Why not `dashboard_tab`? There's already no consistency between tables with `report_` in front and those without, so I would prefer to minimize the already small number of tables with `report_` in front and have most tables not have it.
</blockquote></details>

good point, I think we has been meaning to remove that refix for dashboard, cards...

Will remove

**Review by qnkhuat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29802)* status: 3 of 10 files reviewed, 5 unresolved discussions (waiting on @calherries)

___
*[`enterprise/backend/src/metabase_enterprise/serialization/v2/seed_entity_ids.clj` line 38 at r1](https://reviewable.io/reviews/metabase/metabase/29802#gh-1162447821:-NSjzL3b9U5uYhL6h14x:b-4hqczl) ([raw file](https://github.com/metabase/metabase/blob/a659d80b64a851d1e03e0f8ca95b9ed99e5ca8e6/enterprise/backend/src/metabase_enterprise/serialization/v2/seed_entity_ids.clj#L38)):*
<details><summary><i>Previously, calherries (Cal Herries) wrote…</i></summary><blockquote>

Shouldn't it be `:metabase/model`?
</blockquote></details>

yes it shoud, updated in https://github.com/metabase/metabase/pull/29802/commits/30b923071e46d6a6494b25c3eacd816612b3e26b
___
*[`resources/migrations/000_migrations.yaml` line 14387 at r1](https://reviewable.io/reviews/metabase/metabase/29802#-NSjTWrI5c-Ui-gVa1wF:-NSjzQtB7cnsvRI8Xd59:b3riapm) ([raw file](https://github.com/metabase/metabase/blob/a659d80b64a851d1e03e0f8ca95b9ed99e5ca8e6/resources/migrations/000_migrations.yaml#L14387)):*
<details><summary><i>Previously, qnkhuat (Ngoc Khuat) wrote…</i></summary><blockquote>

good point, I think we has been meaning to remove that refix for dashboard, cards...

Will remove
</blockquote></details>

updated in https://github.com/metabase/metabase/pull/29802/commits/c37161211976ae2de2edfcc6fa9a5e30016ed14d
___
*[`resources/migrations/000_migrations.yaml` line 14450 at r1](https://reviewable.io/reviews/metabase/metabase/29802#gh-1162486993:-NSjzSzsDD-PCwkVSzo9:b-2s1pfc) ([raw file](https://github.com/metabase/metabase/blob/a659d80b64a851d1e03e0f8ca95b9ed99e5ca8e6/resources/migrations/000_migrations.yaml#L14450)):*
<details><summary><i>Previously, qnkhuat (Ngoc Khuat) wrote…</i></summary><blockquote>

I took this because we have `dashboardcard_series.dashboardcard_id` 
</blockquote></details>

changed to `dashboard_tab_id` in https://github.com/metabase/metabase/pull/29802/commits/c37161211976ae2de2edfcc6fa9a5e30016ed14d
___
*[`src/metabase/models/dashboard.clj` line 58 at r1](https://reviewable.io/reviews/metabase/metabase/29802#-NSjUweO5qOG2ngLAw6w:-NSjzWak-qni93pJP7hg:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/a659d80b64a851d1e03e0f8ca95b9ed99e5ca8e6/src/metabase/models/dashboard.clj#L58)):*
<details><summary><i>Previously, calherries (Cal Herries) wrote…</i></summary><blockquote>

Extra whitespace
</blockquote></details>

Done.
___
*[`src/metabase/models/dashboard.clj` line 61 at r1](https://reviewable.io/reviews/metabase/metabase/29802#-NSjV1ez5iLa1GGMAqGw:-NSjzY3VBDhBi6GHrDvG:bvoycxg) ([raw file](https://github.com/metabase/metabase/blob/a659d80b64a851d1e03e0f8ca95b9ed99e5ca8e6/src/metabase/models/dashboard.clj#L61)):*
> ```Clojure
> (mi/define-simple-hydration-method ordered-tabs
>   :ordered_tabs
>   "Return the DashboardCards associated with `dashboard`, in the order they were created."
> ```

https://github.com/metabase/metabase/pull/29802/commits/30b923071e46d6a6494b25c3eacd816612b3e26b

**Review by calherries — APPROVED** [review]
I have a bunch of small suggestions, but no change in behaviour.
:lgtm:

> Reviewed 1 of 10 files at r1, 4 of 5 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29802#-:-NSq-90a6KTHEUd7ZYx1:bckvegw)* status: all files reviewed, 8 unresolved discussions (waiting on @qnkhuat)

___
*[`enterprise/backend/src/metabase_enterprise/serialization/v2/seed_entity_ids.clj` line 38 at r4](https://reviewable.io/reviews/metabase/metabase/29802#-NSptaBQ-HfWunQiwi5l:-NSptaBQ-HfWunQiwi5m:b-wry5eo) ([raw file](https://github.com/metabase/metabase/blob/d7c268a49d0377bf36b7094b4de9cc0c445bbeb1/enterprise/backend/src/metabase_enterprise/serialization/v2/seed_entity_ids.clj#L38)):*
> ```Clojure
>   (into {}
>         (for [model (concat (descendants :toucan1/model) (descendants :metabase/model))
>               :when (mdb.u/toucan-model? model)
> ```

I think this particular call to `models/model?` is redundant, so we can actually remove this `:when` clause. The implementation of `models/model?` is `(isa? model :toucan1/model)`.
___
*[`src/metabase/db/util.clj` line 17 at r4](https://reviewable.io/reviews/metabase/metabase/29802#-NSprrLD9BKDnzpVed0V:-NSprrLD9BKDnzpVed0W:b45h) ([raw file](https://github.com/metabase/metabase/blob/d7c268a49d0377bf36b7094b4de9cc0c445bbeb1/src/metabase/db/util.clj#L17)):*
> ```Clojure
>   (if (keyword? model)
>     true
>     (models/model? model)))
> ```





_Suggestion:_
```Clojure
  (or (keyword? model) (models/model? model)))
```
___
*[`test/metabase/api/dashboard_test.clj` line 1359 at r4](https://reviewable.io/reviews/metabase/metabase/29802#-NSpuPIO6HdFp0u7ZcZH:-NSpuPIO6HdFp0u7ZcZI:b-g1kqjy) ([raw file](https://github.com/metabase/metabase/blob/d7c268a49d0377bf36b7094b4de9cc0c445bbeb1/test/metabase/api/dashboard_test.clj#L1359)):*
> ```Clojure
>                       (t2/select [DashboardCard :size_x :size_y :col :row :parameter_mappings :visualization_settings]
>                                  :dashboard_id dashboard-id)))))))))
> ```

extra whitespace
___
*[`test/metabase/models/dashboard_tab_test.clj` line 45 at r4](https://reviewable.io/reviews/metabase/metabase/29802#-NSpz5ht2a_B--1hfvDQ:-NSpz5ht2a_B--1hfvDR:b10xax5) ([raw file](https://github.com/metabase/metabase/blob/d7c268a49d0377bf36b7094b4de9cc0c445bbeb1/test/metabase/models/dashboard_tab_test.clj#L45)):*
> ```Clojure
> 
>     (testing (str "Check that if a Dashtab of a Dashboard is in a Collection, someone who would not be able to see it under the old "
>                   "artifact-permissions regime will be able to see it if they have permissions for that Collection")
> ```

Maybe I don't know enough about permissions. But is this test redundant, given we've already checked the perms-objects-set is the same as a dashboard's?
___
*[`test/metabase/models/dashboard_tab_test.clj` line 51 at r4](https://reviewable.io/reviews/metabase/metabase/29802#-NSpyZ2f3WA4yPSMvNiv:-NSpyZ2f3WA4yPSMvNiw:b-bm1j39) ([raw file](https://github.com/metabase/metabase/blob/d7c268a49d0377bf36b7094b4de9cc0c445bbeb1/test/metabase/models/dashboard_tab_test.clj#L51)):*
> ```Clojure
>                (mi/can-read? dashtab)))
>         (is (= false
>                (mi/can-write? dashtab)))))
> ```

These are the only tests where the `(= true ...)` expression is on two lines. Why? Either be strict about styling rule having the args on different lines, or don't and at least be consistent and have them all on one line.



_Code quote:_
```Clojure
        (is (= true
               (mi/can-read? dashtab)))
        (is (= false
               (mi/can-write? dashtab)))))
```
___
*[`test/metabase/models/dashboard_tab_test.clj` line 58 at r4](https://reviewable.io/reviews/metabase/metabase/29802#-NSpvbsfEn_8Qfp4C7MJ:-NSpvbsfEn_8Qfp4C7MK:undefined) ([raw file](https://github.com/metabase/metabase/blob/d7c268a49d0377bf36b7094b4de9cc0c445bbeb1/test/metabase/models/dashboard_tab_test.clj#L58)):*
> ```Clojure
>         (is (= true (mi/can-write? dashtab)))))
> 
>     (testing "An user that can't see the Collection that the Dashboard is in can't read and write the dashtab"
> ```





_Suggestion:_
```Clojure
    (testing "A user that can't see the Collection that the Dashboard is in can't read and write the dashtab"
```
___
*[`test/metabase/models/dashboard_tab_test.clj` line 63 at r4](https://reviewable.io/reviews/metabase/metabase/29802#-NSpveUv7pEcN8358oHv:-NSpveUv7pEcN8358oHw:undefined) ([raw file](https://github.com/metabase/metabase/blob/d7c268a49d0377bf36b7094b4de9cc0c445bbeb1/test/metabase/models/dashboard_tab_test.clj#L63)):*
> ```Clojure
>         (is (= false (mi/can-write? dashboard)))))
> 
>     (testing "An user that can see the Collection that the Dashboard is in can read and write the dashtab"
> ```





_Suggestion:_
```Clojure
    (testing "A user that can see the Collection that the Dashboard is in can read and write the dashtab"
```
___
*[`test/metabase/models/dashboard_tab_test.clj` line 81 at r4](https://reviewable.io/reviews/metabase/metabase/29802#-NSpziBc8UzA2t_MddCr:-NSpziBc8UzA2t_MddCs:undefined) ([raw file](https://github.com/metabase/metabase/blob/d7c268a49d0377bf36b7094b4de9cc0c445bbeb1/test/metabase/models/dashboard_tab_test.clj#L81)):*
> ```Clojure
> 
> (deftest hydration-test
>   (testing "hydrate a dashboard will returns all of its tabs"
> ```





_Suggestion:_
```Clojure
  (testing "hydrate a dashboard will return all of its tabs"
```

**Review by qnkhuat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29802)* status: 7 of 11 files reviewed, 8 unresolved discussions (waiting on @calherries)

___
*[`test/metabase/models/dashboard_tab_test.clj` line 45 at r4](https://reviewable.io/reviews/metabase/metabase/29802#-NSpz5ht2a_B--1hfvDQ:-NSsHkq37cBRp6BHHa02:bz2izae) ([raw file](https://github.com/metabase/metabase/blob/d7c268a49d0377bf36b7094b4de9cc0c445bbeb1/test/metabase/models/dashboard_tab_test.clj#L45)):*
<details><summary><i>Previously, calherries (Cal Herries) wrote…</i></summary><blockquote>

Maybe I don't know enough about permissions. But is this test redundant, given we've already checked the perms-objects-set is the same as a dashboard's?
</blockquote></details>

yes but since it's a model on its own so I think it should have perms test on its own, even though dashboard already has a test that checks the same thing. I tend to be more cautious around permission when is possible.
___
*[`test/metabase/models/dashboard_tab_test.clj` line 51 at r4](https://reviewable.io/reviews/metabase/metabase/29802#-NSpyZ2f3WA4yPSMvNiv:-NSsHWRn-QAVFLN98YkD:b-uvkxhi) ([raw file](https://github.com/metabase/metabase/blob/d7c268a49d0377bf36b7094b4de9cc0c445bbeb1/test/metabase/models/dashboard_tab_test.clj#L51)):*
<details><summary><i>Previously, calherries (Cal Herries) wrote…</i></summary><blockquote>

These are the only tests where the `(= true ...)` expression is on two lines. Why? Either be strict about styling rule having the args on different lines, or don't and at least be consistent and have them all on one line.
</blockquote></details>

agree, will make it all in one line.
___
*[`test/metabase/models/dashboard_tab_test.clj` line 58 at r4](https://reviewable.io/reviews/metabase/metabase/29802#-NSpvbsfEn_8Qfp4C7MJ:-NSsIg7VDtBP19cyVjB3:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/d7c268a49d0377bf36b7094b4de9cc0c445bbeb1/test/metabase/models/dashboard_tab_test.clj#L58)):*
> ```Clojure
>         (is (= true (mi/can-write? dashtab)))))
> 
>     (testing "An user that can't see the Collection that the Dashboard is in can't read and write the dashtab"
> ```

Done.
___
*[`test/metabase/models/dashboard_tab_test.clj` line 63 at r4](https://reviewable.io/reviews/metabase/metabase/29802#-NSpveUv7pEcN8358oHv:-NSsIgXT0lKa_D4Y_E4i:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/d7c268a49d0377bf36b7094b4de9cc0c445bbeb1/test/metabase/models/dashboard_tab_test.clj#L63)):*
> ```Clojure
>         (is (= false (mi/can-write? dashboard)))))
> 
>     (testing "An user that can see the Collection that the Dashboard is in can read and write the dashtab"
> ```

Done.
___
*[`test/metabase/models/dashboard_tab_test.clj` line 81 at r4](https://reviewable.io/reviews/metabase/metabase/29802#-NSpziBc8UzA2t_MddCr:-NSsIh86APTCwW4fcem1:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/d7c268a49d0377bf36b7094b4de9cc0c445bbeb1/test/metabase/models/dashboard_tab_test.clj#L81)):*
> ```Clojure
> 
> (deftest hydration-test
>   (testing "hydrate a dashboard will returns all of its tabs"
> ```

Done.

**Review by calherries — APPROVED** [review]
> Reviewed 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29802)* status: all files reviewed, 1 unresolved discussion (waiting on @qnkhuat)

**Pull request merged**